Saving arrays from np.broadcast_arrays fails

This:
```python
def test_broadcasting(tmpdir):
   attrs = np.broadcast_arrays(np.array([10,20]), np.array(10), np.array(10))
   tree = {'one': attrs[0],
           'two': attrs[1],
           'three': attrs[2]}
   assert_roundtrip_tree(tree, tmpdir)
```

Fails with a 
```
E           jsonschema.exceptions.ValidationError: {'shape': [2], 'source': 1, 'datatype': 'int64', 'byteorder': 'little', 'strides': [0]} is not valid under any of the given schemas
```

(full traceback: https://gist.github.com/Cadair/9ff970a06c4f90bf409a86b4022505f8)

I am assuming it's a problem with 'strides'

--- a/CHANGES.rst
+++ b/CHANGES.rst
@@ -3,6 +3,11 @@
 
 - Add API function for retrieving history entries. [#501]
 
+2.0.2 (unreleased)
+------------------
+
+- Allow serialization of broadcasted ``numpy`` arrays. [#507]
+
 2.0.1 (2018-05-08)
 ------------------
 

--- a/asdf/generic_io.py
+++ b/asdf/generic_io.py
@@ -372,7 +372,7 @@ class GenericFile(object):
     """
 
     def write_array(self, array):
-        _array_tofile(None, self.write, array)
+        _array_tofile(None, self.write, np.ascontiguousarray(array))
 
     def seek(self, offset, whence=0):
         """

--- a/asdf/generic_io.py
+++ b/asdf/generic_io.py
@@ -751,7 +751,7 @@ class RealFile(RandomAccessFile):
             arr.flush()
             self.fast_forward(len(arr.data))
         else:
-            _array_tofile(self._fd, self._fd.write, arr)
+            _array_tofile(self._fd, self._fd.write, np.ascontiguousarray(arr))
 
     def can_memmap(self):
         return True

--- a/asdf/tags/core/ndarray.py
+++ b/asdf/tags/core/ndarray.py
@@ -386,14 +386,19 @@ class NDArrayType(AsdfType):
     @classmethod
     def to_tree(cls, data, ctx):
         base = util.get_array_base(data)
-        block = ctx.blocks.find_or_create_block_for_array(data, ctx)
         shape = data.shape
         dtype = data.dtype
         offset = data.ctypes.data - base.ctypes.data
-        if data.flags[b'C_CONTIGUOUS']:
-            strides = None
-        else:
-            strides = data.strides
+        strides = None
+
+        if not data.flags.c_contiguous:
+            # We do not want to encode strides for broadcasted arrays
+            if not all(data.strides):
+                data = np.ascontiguousarray(data)
+            else:
+                strides = data.strides
+
+        block = ctx.blocks.find_or_create_block_for_array(data, ctx)
 
         result = {}
 

--- a/asdf/tags/core/tests/test_ndarray.py
+++ b/asdf/tags/core/tests/test_ndarray.py
@@ -803,3 +803,9 @@ def test_tagged_object_array(tmpdir):
         objdata.flat[i] = Quantity(i, 'angstrom')
 
     helpers.assert_roundtrip_tree({'bizbaz': objdata}, tmpdir)
+
+
+def test_broadcasted_array(tmpdir):
+    attrs = np.broadcast_arrays(np.array([10,20]), np.array(10), np.array(10))
+    tree = {'one': attrs[1] }#, 'two': attrs[1], 'three': attrs[2]}
+    helpers.assert_roundtrip_tree(tree, tmpdir)

